File: chad_c/Algorithms/merge_sort.c
#include <stddef.h>

void merge(int arr[], int temp[], size_t left, size_t mid, size_t right);

void merge_sort(int arr[], int temp[], size_t left, size_t right)
{
    if (left >= right)
    {
        return;
    }

    size_t mid = (left + right) / 2;

    merge_sort(arr, temp, left, mid);
    merge_sort(arr, temp, mid + 1, right);

    merge(arr, temp, left, mid, right);
}

void merge(int arr[], int temp[], size_t left, size_t mid, size_t right)
{
    size_t i = left;
    size_t j = mid + 1;
    size_t k = left;

    while (i <= mid && j <= right)
    {
        if (arr[i] <= arr[j])
        {
            temp[k++] = arr[i++];
        }

        else
        {
            temp[k++] = arr[j++];
        }
    }

    while (i <= mid)
    {
        temp[k++] = arr[i++];
    }

    while (j <= right)
    {
        temp[k++] = arr[j++];
    }

    for (size_t p = left; p <= right; p++)
    {
        arr[p] = temp[p];
    }
}

File: virgin_python/Algorithms/merge_sort.py
def merge_sort(arr):

    if len(arr) <= 1: # Base case
        return arr
    
    # Slice in two
    mid = len(arr) // 2
    left = arr[:mid]
    right = arr[mid:]

    # Keep calling the function until the base case is reached
    left = merge_sort(left)
    right = merge_sort(right)

    # When arr slices are of length 1, merge the sorted arr up
    return merge(left, right)


def merge(left, right):

    l = [] # Merged list
    i = 0 # Left arr pointer
    j = 0 # Right arr pointer

    while (i < len(left) and j < len(right)):

        if left[i] <= right[j]:
            l.append(left[i])
            i += 1

        else:
            l.append(right[j])
            j += 1

    
    # Clean up code
    while i < len(left):
        l.append(left[i])
        i += 1

    while j < len(right):
        l.append(right[j])
        j += 1

    # Return the merged list back to the caller
    return l


# Time:
#   O(n logn) => Slicing and then merging are logn, but sorting is n

# Space:
#   O(n) => Extra space is needed to store the sorted sub-arr's